Common/Securities/Cash.cs

```diff
@@ -156,105 +204,133 @@ SecurityChanges changes
             if (markets.ContainsKey(SecurityType.Cfd) && !markets.ContainsKey(SecurityType.Forex))
             {
                 markets.Add(SecurityType.Forex, markets[SecurityType.Cfd]);
             }
+
             if (markets.ContainsKey(SecurityType.Forex) && !markets.ContainsKey(SecurityType.Cfd))
             {
                 markets.Add(SecurityType.Cfd, markets[SecurityType.Forex]);
             }
 
+            var minimumResolution = subscriptions.Subscriptions.Select(x => x.Resolution).DefaultIfEmpty(Resolution.Minute).Min();
+
+            var conversionSecuritiesList = new List<ConversionSecurity>();
+
+            var requiredSecurities = new List<Security>();
+
+            // return needed pairs for full conversion from one currency to another
+            var pathProvider = new CurrencyConversion.PathProvider.LinearMax2SearchProvider();
+
             var potentials = Currencies.CurrencyPairs.Select(fx => CreateSymbol(marketMap, fx, markets, SecurityType.Forex))
-                .Concat(Currencies.CfdCurrencyPairs.Select(cfd => CreateSymbol(marketMap, cfd, markets, SecurityType.Cfd)))
-                .Concat(Currencies.CryptoCurrencyPairs.Select(crypto => CreateSymbol(marketMap, crypto, markets, SecurityType.Crypto)));
+              .Concat(Currencies.CfdCurrencyPairs.Select(cfd => CreateSymbol(marketMap, cfd, markets, SecurityType.Cfd)))
+              .Concat(Currencies.CryptoCurrencyPairs.Select(crypto => CreateSymbol(marketMap, crypto, markets, SecurityType.Crypto)));
 
-            var minimumResolution = subscriptions.Subscriptions.Select(x => x.Resolution).DefaultIfEmpty(Resolution.Minute).Min();
 
-            foreach (var symbol in potentials)
+            // add securities symbols from securitiesToSearch collection (from Currencies.cs)
+            foreach (var knownSecurity in securitiesToSearch)
+            {
+                pathProvider.AddEdge(knownSecurity.Symbol.Value, knownSecurity.Type);
+            }
+
+            // add custom securities
+            foreach (var potential in potentials)
             {
-                if (symbol.Value == normal || symbol.Value == invert)
+                pathProvider.AddEdge(potential.Value, potential.SecurityType);
+            }
+
+            // calculate conversion path
+            var shortestPath = pathProvider.FindShortestPath(Symbol, CashBook.AccountCurrency);
+
+            // for each step, find existing security, and if it doesn't exist, make new one
+            // also build ConversionRateSecurity list
+            foreach (var step in shortestPath.Steps)
+            {
+                Security security = null;
+
+                var existingSecuritySet = securitiesToSearch.Where(s => s.Symbol.Value == step.Edge.PairSymbol);
+
+                if (existingSecuritySet.Any())
+                {
+                    security = existingSecuritySet.Single();
```

Comment:
Isn't algorithm only per one exchange?

Or algorithm can be subscribed to many (non-ticker) symbols from different exchanges?

Reply:
Ok, I got it.. algorithm trades only on one exchange, but can subscribe to many symbols which can be from different exchanges. We can solve this problem last, no?

Because, there should be preference of choosing to use data from same exchange if possible.
Or again, volume based, but I don't know how to get 24h volume average from symbol.

Reply:
Let's keep it simple @viliwonka =) LEAN supports 1 exchange today. Let's start with the assumption of a single exchange. Data subscription issues aren't a concern of the currency conversion logic.

Reply:
while it's true that we *currently* only support trading against a single brokerage, the LEAN engine has absolutely no issues receiving data from multiple exchanges, and therefore, it's possible to have multiple currency pairs.

If we keep the `Single` call, than *all* algorithms adding `EURUSD` from FXCM and Oanda (even if only paper trading), will throw an exception on initialization.

Reply:
No worries - I am focusing on implementing `SecurityCurrencyConversion`, we will focus on this multi-exchange issue later, if neede.